Add Swift Code Generator

rogerluan opened the pull request:
## Summary

This PR adds a new Swift Code Generator that uses Swift 5.2.
Resolves #1368

## Tests

<details><summary>Click to see the tests results</summary>
<p>

## isSubtituting = true, isGlobal = true

Pattern: `(http[s]?(\S*))(.*)`

Substitution String: `[$1]($1?ref=rogeroba)$3`

<details><summary>Swift Output</summary>
<p>

```swift
import Foundation

let pattern = #"(http[s]?(\S*))(.*)"#
let regex = try! NSRegularExpression(pattern: pattern, options: [])
let testString = #"""
Sample markdown string that contains a link: http://github.com/rogerluan/

*GitHub Profile*

1. https://github.com/rogerluan/ - This is my profile

https://twitter.com/rogerluan_
"""#
let stringRange = NSRange(location: 0, length: testString.utf16.count)
let substitutionString = #"[$1]($1?ref=rogeroba)$3"#
let result = regex.stringByReplacingMatches(in: testString, range: stringRange, withTemplate: substitutionString)
print(result)
```

</p>
</details>

<details><summary>Swift Script Execution Result</summary>
<p>

```
Sample markdown string that contains a link: [http://github.com/rogerluan/](http://github.com/rogerluan/?ref=rogeroba)

*GitHub Profile*

1. [https://github.com/rogerluan/](https://github.com/rogerluan/?ref=rogeroba) - This is my profile

[https://twitter.com/rogerluan_](https://twitter.com/rogerluan_?ref=rogeroba)
```

</p>
</details>

## isSubtituting = true, isGlobal = false

Pattern: `(http[s]?(\S*))(.*)`

Substitution String: `[$1]($1?ref=rogeroba)$3`

<details><summary>Swift Output</summary>
<p>

```swift
import Foundation

let pattern = #"(http[s]?(\S*))(.*)"#
let regex = try! NSRegularExpression(pattern: pattern, options: [])
let testString = #"""
Sample markdown string that contains a link: http://github.com/rogerluan/

*GitHub Profile*

1. https://github.com/rogerluan/ - This is my profile

https://twitter.com/rogerluan_
"""#
let stringRange = NSRange(location: 0, length: testString.utf16.count)
let substitutionString = #"[$1]($1?ref=rogeroba)$3"#
let lookupRange = (testString as NSString).range(of: pattern, options: .regularExpression, range: stringRange)
if lookupRange.intersection(stringRange) != nil {
    let result = regex.stringByReplacingMatches(in: testString, range: lookupRange, withTemplate: substitutionString)
    print(result)
} else {
    print("No matches were found.")
}
```

</p>
</details>

<details><summary>Swift Script Execution Result</summary>
<p>

```
Sample markdown string that contains a link: [http://github.com/rogerluan/](http://github.com/rogerluan/?ref=rogeroba)

*GitHub Profile*

1. https://github.com/rogerluan/ - This is my profile

https://twitter.com/rogerluan_
```

</p>
</details>

## isSubtituting = false, isGlobal = true

Pattern: `(http[s]?(\S*))(.*)`

Substitution String: `[$1]($1?ref=rogeroba)$3`

<details><summary>Swift Output</summary>
<p>

```swift
import Foundation

let pattern = #"(http[s]?(\S*))(.*)"#
let regex = try! NSRegularExpression(pattern: pattern, options: [])
let testString = #"""
Sample markdown string that contains a link: http://github.com/rogerluan/

*GitHub Profile*

1. https://github.com/rogerluan/ - This is my profile

https://twitter.com/rogerluan_
"""#
let stringRange = NSRange(location: 0, length: testString.utf16.count)
let matches = regex.matches(in: testString, range: stringRange)
var result: [[String]] = []
for match in matches {
    var groups: [String] = []
    for rangeIndex in 1 ..< match.numberOfRanges {
        groups.append((testString as NSString).substring(with: match.range(at: rangeIndex)))
    }
    if !groups.isEmpty {
        result.append(groups)
    }
}
print(result)
```

</p>
</details>

<details><summary>Swift Script Execution Result</summary>
<p>

```
[["http://github.com/rogerluan/", "://github.com/rogerluan/", ""], ["https://github.com/rogerluan/", "://github.com/rogerluan/", " - This is my profile"], ["https://twitter.com/rogerluan_", "://twitter.com/rogerluan_", ""]]
```

</p>
</details>

## isSubtituting = false, isGlobal = false

Pattern: `(http[s]?(\S*))(.*)`

Substitution String: `[$1]($1?ref=rogeroba)$3`

<details><summary>Swift Output</summary>
<p>

```swift
import Foundation

let pattern = #"(http[s]?(\S*))(.*)"#
let regex = try! NSRegularExpression(pattern: pattern, options: [])
let testString = #"""
Sample markdown string that contains a link: http://github.com/rogerluan/

*GitHub Profile*

1. https://github.com/rogerluan/ - This is my profile

https://twitter.com/rogerluan_
"""#
let stringRange = NSRange(location: 0, length: testString.utf16.count)
if let firstMatch = regex.firstMatch(in: testString, range: stringRange) {
    let result: [String] = (1 ..< firstMatch.numberOfRanges).map { (testString as NSString).substring(with: firstMatch.range(at: $0)) }
    print(result)
} else {
    print("No matches were found.")
}
```

</p>
</details>

<details><summary>Swift Script Execution Result</summary>
<p>

```
["http://github.com/rogerluan/", "://github.com/rogerluan/", ""]
```

</p>
</details>

## isSubtituting = true, isGlobal = true, includes "x" flag

Pattern: `^(http[s]?(\S*))(.*)`

Substitution String: `[$1]($1?ref=rogeroba)$3`

<details><summary>Swift Output</summary>
<p>

```swift
import Foundation

let pattern = #"^(http[s]?(\S*))(.*)"#
let regex = try! NSRegularExpression(pattern: pattern, options: .anchorsMatchLines)
let testString = #"""
Sample markdown string that contains a link: http://github.com/rogerluan/

*GitHub Profile*

1. https://github.com/rogerluan/ - This is my profile

https://twitter.com/rogerluan_
"""#
let stringRange = NSRange(location: 0, length: testString.utf16.count)
let substitutionString = #"[$1]($1?ref=rogeroba)$3"#
let result = regex.stringByReplacingMatches(in: testString, range: stringRange, withTemplate: substitutionString)
print(result)
```

</p>
</details>

<details><summary>Swift Script Execution Result</summary>
<p>

```
Sample markdown string that contains a link: http://github.com/rogerluan/

*GitHub Profile*

1. https://github.com/rogerluan/ - This is my profile

[https://twitter.com/rogerluan_](https://twitter.com/rogerluan_?ref=rogeroba)
```

</p>
</details>

## isSubtituting = true, isGlobal = true, doesn't include "x" flag

Pattern: `^(http[s]?(\S*))(.*)`

Substitution String: `[$1]($1?ref=rogeroba)$3`

<details><summary>Swift Output</summary>
<p>

```swift
import Foundation

let pattern = #"^(http[s]?(\S*))(.*)"#
let regex = try! NSRegularExpression(pattern: pattern, options: [])
let testString = #"""
Sample markdown string that contains a link: http://github.com/rogerluan/

*GitHub Profile*

1. https://github.com/rogerluan/ - This is my profile

https://twitter.com/rogerluan_
"""#
let stringRange = NSRange(location: 0, length: testString.utf16.count)
let substitutionString = #"[$1]($1?ref=rogeroba)$3"#
let result = regex.stringByReplacingMatches(in: testString, range: stringRange, withTemplate: substitutionString)
print(result)
```

</p>
</details>

<details><summary>Swift Script Execution Result</summary>
<p>

```
Sample markdown string that contains a link: http://github.com/rogerluan/

*GitHub Profile*

1. https://github.com/rogerluan/ - This is my profile

https://twitter.com/rogerluan_
```

</p>
</details>

## Test using all flags

Tested using: `const flags = ["g", "m", "i", "x", "s", "X", "u", "J", "A", "U", "D"]`

<details><summary>Swift Output</summary>
<p>

```swift
import Foundation

// WARNING: You included a flag that Swift doesn't support: X
//          When this flag is set, it causes any backslash in the pattern that is followed by a letter that has no special meaning to cause an error.
//          By default in Swift, a backslash followed by a letter with no special meaning will be treated as a literal.
// WARNING: You included a flag that Swift doesn't support: u
//          When this flag is set, it makes the pattern and subject strings to be treated as unicode.
//          Swift already treats the pattern and subject strings as unicode by default, so including this flag is redundant.
// WARNING: You included a flag that Swift doesn't support: J
//          Wehn this flag is set, it allows duplicated capturing group names.
//          By default, Swift captures only the last value matched for a repeated capture group.
//          As an alternative, the pattern can be modified to contain one capturing group per group you want to get in the result.
// WARNING: You included a flag that Swift doesn't support: A
//          When this flag is set, it causes the pattern to be "anchored", that is, it is constrained to match only at the start of the string which is being searched (the "subject string").
//          As an alternative, this effect can also be achieved by appropriate constructs in the pattern itself.
// WARNING: You included a flag that Swift doesn't support: U
//          When this flag is set, it inverts the "greediness" of the quantifiers so that they are not greedy by default, but become greedy if followed by '?'.
//          As an alternative, this effect can also be achieved by setting a (?U) modifier setting within the pattern or by a question mark behind a quantifier (e.g. .*?).
// WARNING: You included a flag that Swift doesn't support: D
//          When this flag is set, it forces the a dollar sign ('$'), to always match end of the string, instead of end of the line. This option is ignored if the 'm' flag is set.
//          There's no equivalent implementation of this flag in Swift, but you may want to check the 'm' flag instead.

let pattern = #"^(http[s]?(\S*))(.*)"#
let regex = try! NSRegularExpression(pattern: pattern, options: [.anchorsMatchLines, .caseInsensitive, .allowCommentsAndWhitespace, .dotMatchesLineSeparators])
let testString = #"""
Sample markdown string that contains a link: http://github.com/rogerluan/

*GitHub Profile*

1. https://github.com/rogerluan/ - This is my profile

https://twitter.com/rogerluan_
"""#
let stringRange = NSRange(location: 0, length: testString.utf16.count)
let substitutionString = #"[$1]($1?ref=rogeroba)$3"#
let result = regex.stringByReplacingMatches(in: testString, range: stringRange, withTemplate: substitutionString)
print(result)
```

</p>
</details>

As you can see, I added warning comments when the user attempts to use unsupported flags, with some helpful explanation, default behavior, and alternatives.

</p>
</details>

## Design Decisions

- I opted to not add redundant trailing semicolons. Let me know if this is not desired and I can change this style.
- I didn't implement support to all the custom flags, because some of them are not supported by Swift's `NSRegularExpression`.
  - Supported flags: `g`, `m`, `i`, `x`, `s`.
  - Unsupported flags: `X`, `u`, `J`, `A`, `U`, `D` - when a user attempts to use an unsupported flag, a warning will be put in the Swift code output, as you can observe in the tests shown above.

## Next Steps

There're some questions I had that I couldn't figure out on my own.

1. What is the `delimiter` constant, and what should I use it for? I didn't see any use for it in my examples, but maybe it's because I don't understand it.
2. Since there's no test suite, I tested everything by hardcoding consts in the script. The regex const crashes on `_escapeSwiftString(input)` because it's not a String. I can't express the regex const as a String because then some components can't be escaped (e.g. `\S` in my examples). To work around this, I added a `if (input.match(/^\/(.*)\/$/g)) input = input.slice(1, -1)` (remove leading and trailing `/`) while testing. Not sure if this code needs to be present in the production version of the script, because I'm uncertain about the type of the `regex` prop in production 🤔 

## Feedback

Using the `CodeString` indentation was quite confusing 😅 Initially I thought we had to specify the indentation in "level" rather than "number of spaces", so I replaced 1 with 4, 2 with 8, etc. I also thought that the specified indentation would be used in that very same line, and actually it is used only on the next line onwards (so we have to add the indentation a line before, e.g. when opening `{`).

This is quite confusing, and the end result is extremely hard to be read. I'd much rather do e.g.:

```js
codeString.append(`let matches = regex.matches(in: testString, range: stringRange)`)
codeString.append(`var result: [[String]] = []`)
codeString.append(`for match in matches {`)
codeString.append(`    var groups: [String] = []`)
codeString.append(`    for rangeIndex in 1 ..< match.numberOfRanges {`)
codeString.append(`        groups.append((testString as NSString).substring(with: match.range(at: rangeIndex)))`)
codeString.append(`    }`)
codeString.append(`    if !groups.isEmpty {`)
codeString.append(`        result.append(groups)`)
codeString.append(`    }`)
codeString.append(`}`)
codeString.append(`print(result)`)
```

Than:

```js
codeString.append(`let matches = regex.matches(in: testString, range: stringRange)`)
codeString.append(`var result: [[String]] = []`)
codeString.append(`for match in matches {`, 4)
codeString.append(`var groups: [String] = []`)
codeString.append(`for rangeIndex in 1 ..< match.numberOfRanges {`, 8)
codeString.append(`groups.append((testString as NSString).substring(with: match.range(at: rangeIndex)))`, 4)
codeString.append(`}`)
codeString.append(`if !groups.isEmpty {`, 8)
codeString.append(`result.append(groups)`, 4)
codeString.append(`}`, 0)
codeString.append(`}`)
codeString.append(`print(result)`)
```

As the first option is way easier to be read imo 🙇 But, ultimately, maybe using full multiline string literals would be even more readable.
Anyways, just a quick feedback about the API :) I still sticked to using the `CodeString` as designed , because I believe that's how you'd rather see this file.

rogerluan commented on Swift5_2.js:
```diff
@@ -0,0 +1,111 @@
+        codeString.append(`}`)
+      }
+    } else {
+        if (isGlobal) {
+          codeString.append(`let matches = regex.matches(in: testString, range: stringRange)`)
+          codeString.append(`var result: [[String]] = []`)
+          codeString.append(`for match in matches {`, 4)
+          codeString.append(`var groups: [String] = []`)
+          codeString.append(`for rangeIndex in 1 ..< match.numberOfRanges {`, 8)
+          codeString.append(`groups.append((testString as NSString).substring(with: match.range(at: rangeIndex)))`, 4)
+          codeString.append(`}`)
+          codeString.append(`if !groups.isEmpty {`, 8)
+          codeString.append(`result.append(groups)`, 4)
+          codeString.append(`}`, 0)
+          codeString.append(`}`)
+          codeString.append(`print(result)`)
+        } else {
+          codeString.append(`if let firstMatch = regex.firstMatch(in: testString, range: stringRange) {`, 4)
+          codeString.append(`let result: [String] = (1 ..< firstMatch.numberOfRanges).map { (testString as NSString).substring(with: firstMatch.range(at: $0)) }`)
+          codeString.append(`print(result)`, 0)
+          codeString.append(`} else {`, 4)
+          codeString.append(`print("No matches were found.")`, 0)
+          codeString.append(`}`)
+        }
```
Here, if I were to follow the PHP template available on GitHub, as well as the other code generators I saw live on the website, we'd end up dumping every "match", something like this:

```
- <NSExtendedRegularExpressionCheckingResult: 0x600000b38f80>{4, 43}{<NSRegularExpression: 0x600002b1fa20> (http[s]?(\S*))(.*) 0x0} #0
  - super: NSRegularExpressionCheckingResult
    - super: NSTextCheckingResult
      - super: NSObject
- <NSExtendedRegularExpressionCheckingResult: 0x600000b39580>{66, 28}{<NSRegularExpression: 0x600002b1fa20> (http[s]?(\S*))(.*) 0x0} #0
  - super: NSRegularExpressionCheckingResult
    - super: NSTextCheckingResult
      - super: NSObject
- <NSExtendedRegularExpressionCheckingResult: 0x600000b3a280>{99, 18}{<NSRegularExpression: 0x600002b1fa20> (http[s]?(\S*))(.*) 0x0} #0
  - super: NSRegularExpressionCheckingResult
    - super: NSTextCheckingResult
      - super: NSObject
```

I don't think the dump above is particularly useful, and oftentimes users won't know how to interpret the match object (`NSTextCheckingResult`). For this reason, I propose printing an array of matching groups (`[[String]]`). This array is more useful and it's probably closer to what users are lookin for.

If this is **not** desired, the highlighted code can be replaced with the code below, that dumps the `NSTextCheckingResult ` objects:

```suggestion
        if (isGlobal) {
          codeString.append(`let matches = regex.matches(in: testString, range: stringRange)`)
          codeString.append(`for match in matches {`, 4)
          codeString.append(`dump(match)`, 0)
          codeString.append(`}`)
        } else {
          codeString.append(`if let firstMatch = regex.firstMatch(in: testString, range: stringRange) {`, 4)
          codeString.append(`print(match)`, 0)
          codeString.append(`} else {`, 4)
          codeString.append(`print("No matches were found.")`, 0)
          codeString.append(`}`)
        }
```

firasdib replied:
Seems fine as is!

Doqnach commented:
> * I didn't implement support to all the custom flags, only `x`, leaving all these flags unsupported: `g, m, i, X, s, u, U, a, J, D` (`g` flag is ignored on this script, although I do parse the `isGlobal` variable 👍 ). Support to other flags may come in a follow up PR.

It is rather important that all flags that the language supports are supported by the code generator. If it wouldn't, issues would start to pile up in github for people who run into issues with it which would give them a less good experience with the website.
If the language doesn't support certain flags it's of course a different story (and unsure on how to deal with that then actually?).

rogerluan commented:
> 
> 
> > * I didn't implement support to all the custom flags, only `x`, leaving all these flags unsupported: `g, m, i, X, s, u, U, a, J, D` (`g` flag is ignored on this script, although I do parse the `isGlobal` variable 👍 ). Support to other flags may come in a follow up PR.
> 
> 
> 
> It is rather important that all flags that the language supports are supported by the code generator. If it wouldn't, issues would start to pile up in github for people who run into issues with it which would give them a less good experience with the website.
> 
> If the language doesn't support certain flags it's of course a different story (and unsure on how to deal with that then actually?).
> 
> 

Thanks for your review!
I checked all the currently supported languages and several don't support custom flags either (_apparently_), or not all of the flags. I just took a quick glance though - as I'm not proficient in all the supported languages, I might've also misinterpreted its code.

Regex flags in Swift are often not simple parameters/options in the language, though, unfortunately. See how the `g` flag (represented as `isGlobal` variable) take very  distinct approaches, for instance. If each of the 11 flags require an if-else statement, we'd end up with 121 cases 🤓hence why I deferred this implementation to be better investigated later.

You have a point though, I'll look into the details of implementing support for all flags and will update this PR with comments when I do so 👍

rogerluan commented:
I just added support to the flags that exist in Swift's `NSRegularExpression`.

<details><summary>Click to see an output example showing the changes made</summary>
<p>

```swift
import Foundation

// WARNING: You included a flag that Swift doesn't support: X
//          When this flag is set, it causes any backslash in the pattern that is followed by a letter that has no special meaning to cause an error.
//          By default in Swift, a backslash followed by a letter with no special meaning will be treated as a literal.
// WARNING: You included a flag that Swift doesn't support: u
//          When this flag is set, it makes the pattern and subject strings to be treated as unicode.
//          Swift already treats the pattern and subject strings as unicode by default, so including this flag is redundant.
// WARNING: You included a flag that Swift doesn't support: J
//          Wehn this flag is set, it allows duplicated capturing group names.
//          By default, Swift captures only the last value matched for a repeated capture group.
//          As an alternative, the pattern can be modified to contain one capturing group per group you want to get in the result.
// WARNING: You included a flag that Swift doesn't support: A
//          When this flag is set, it causes the pattern to be "anchored", that is, it is constrained to match only at the start of the string which is being searched (the "subject string").
//          As an alternative, this effect can also be achieved by appropriate constructs in the pattern itself.
// WARNING: You included a flag that Swift doesn't support: U
//          When this flag is set, it inverts the "greediness" of the quantifiers so that they are not greedy by default, but become greedy if followed by '?'.
//          As an alternative, this effect can also be achieved by setting a (?U) modifier setting within the pattern or by a question mark behind a quantifier (e.g. .*?).
// WARNING: You included a flag that Swift doesn't support: D
//          When this flag is set, it forces the a dollar sign ('$'), to always match end of the string, instead of end of the line. This option is ignored if the 'm' flag is set.
//          There's no equivalent implementation of this flag in Swift, but you may want to check the 'm' flag instead.

let pattern = #"^(http[s]?(\S*))(.*)"#
let regex = try! NSRegularExpression(pattern: pattern, options: [.anchorsMatchLines, .caseInsensitive, .allowCommentsAndWhitespace, .dotMatchesLineSeparators])
let testString = #"""
Sample markdown string that contains a link: http://github.com/rogerluan/

*GitHub Profile*

1. https://github.com/rogerluan/ - This is my profile

https://twitter.com/rogerluan_
"""#
let stringRange = NSRange(location: 0, length: testString.utf16.count)
let substitutionString = #"[$1]($1?ref=rogeroba)$3"#
let result = regex.stringByReplacingMatches(in: testString, range: stringRange, withTemplate: substitutionString)
print(result)
```

</p>
</details>

I used these flags to generate the output below: `const flags = ["g", "m", "i", "x", "s", "X", "u", "J", "A", "U", "D"]`

As you can see, I added warning comments when the user attempts to use unsupported flags, with some helpful explanation, default behavior, and alternatives.

@Doqnach @firasdib looking forward to hear your thoughts on this new generator! 🙏

firasdib commented:
You just need to add support for the flags that your language supports. The rest should be filtered out, in my opinion. If you have an alternate flag name for a given flag, it should be mapped to that. I'll look through this in more detail in the coming days!

rogerluan commented:
Yup, that's exactly what I did :) Mapped the supported flag to the language's regex options :) 
Looking forward to it! 🚀 thanks @firasdib

firasdib commented:
Looks great! Minor changes to your flags code, rest seems fine. I'll include this in the next release!

rogerluan commented:
Awesome! I just saw it live in regex101.com 🎉 Thanks @firasdib again for providing such an amazing regex editor and debugger 🚀

romdevios commented on Swift5_2.js:
```diff
@@ -0,0 +1,111 @@
+    let isMultiline = input.match(/\n/g)
+    // Wrap the string literal around # for each # found in the input string, plus 1 extra #
+    var hashes = "#"
+    while (input.search("#") != -1) {
```
Why did you replace `input.search("hashes")` with `input.search("#")`?
Now if you type `##` and try to generate swift code, you will end up in an infinite loop.
This logic should add as many hashes as the maximum number of them in a row appears in a string.
For example `I'm using # and ## for this line` should be escaped as `###"I'm using # and ## for this line"###`

rogerluan replied:
Good catch! I initially thought the loop would iterate through the string character by character once, but I see it's not what it's doing now 😅 

Thanks for that! 🙌

romdevios commented:
Please fix the infinity loop!

Why did you replace `input.search("hashes")` with `input.search("#")` in escaping function?
Now if you type `##` and try to generate swift code, you will end up in an infinite loop.
This logic should add as many hashes as the maximum number of them in a row appears in a string.
For example, `I'm using # and ## for this line` should be escaped as `###"I'm using # and ## for this line"###`

firasdib commented:
Issue should be resolved now.

rogerluan commented on Swift5_2.js:
```diff
@@ -0,0 +1,183 @@
+          codeString.append(`for match in matches {`, 4)
+          codeString.append(`var groups: [String] = []`)
+          codeString.append(`for rangeIndex in 1 ..< match.numberOfRanges {`, 8)
+          codeString.append(`groups.append((testString as NSString).substring(with: match.range(at: rangeIndex)))`, 4)
```
```suggestion
          codeString.append(`guard !NSEqualRanges(nsRange, NSMakeRange(NSNotFound, 0)) else { continue }`, 4)
          codeString.append(`let string = (testString as NSString).substring(with: match.range(at: rangeIndex))`, 4)
          codeString.append(`groups.append(string)`, 4)
```
